fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -586,6 +594,123 @@ private Tuple recordVersionKey(@Nonnull Tuple primaryKey) {
         }
     }
 
+    @API(API.Status.EXPERIMENTAL)
+    private <M extends Message> void updateSyntheticIndexes(@Nullable FDBStoredRecord<M> oldRecord,
+                                                            @Nullable FDBStoredRecord<M> newRecord,
+                                                            @Nonnull final List<CompletableFuture<Void>> futures) {
+        final SyntheticRecordPlanner planner = new SyntheticRecordPlanner(this);
+        final int pipelineSize = getPipelineSize(PipelineOperation.IN_JOIN);
+        if (oldRecord != null && newRecord != null && oldRecord.getRecordType() == newRecord.getRecordType()) {
+            // TODO: An important optimization here is determining that no field used in the join condition or
+            //  indexed in the synthetic record is changed, in which case all this can be skipped.
+            final SyntheticRecordFromStoredRecordPlan plan = planner.fromStoredType(newRecord.getRecordType(), true);
+            if (plan == null) {
+                return;
+            }
+            final Map<RecordType, Collection<IndexMaintainer>> maintainers = getSyntheticMaintainers(plan.getSyntheticRecordTypes());
+            final Map<Tuple, FDBSyntheticRecord> oldRecords = new HashMap<>();
```

Comment:
If the index maintainers have thread safety issues, I think this does as well, though I suppose that that's up for some discussion right now.

Reply:
The safe thing to do then is probably to use `forEach` instead of `forEachAsync`.

Reply:
Or `forEachAsync` with `pipelineSize = 1`.